// goodreads/node_modules/get-shelves.js
var https = require("https");

var config = {
	host: "www.goodreads.com",
	port: 443,
	key: "n8hATXd6wzepLEZo5zlhkg",
	user: "4175880-retorick"
};

exports.getShelves = function (req, res) {
	shelvesHandler(res).end();
}

function shelvesHandler(res) {
	var options = {
		hostname: config.host,
		port: config.port,
		path: "/shelf/list.json?user_id=" + config.user + "&key=" + config.key,
		method: "GET"
	};

	var handler = 
		https.request(options, function(httpRes) {
			var body = "";
			httpRes.on("data", function(piece) {
				body += piece;
			});

			httpRes.on("end", function() {
				var json = JSON.parse(body);
				var shelves = json.shelves;
				var shelfList = [];
				shelves.user_shelves.forEach (function(obj) {
					var shelfObj = getShelfObj(obj);
					shelfList.push(shelfObj);
				});

				res.json(shelfList);
			});
		});

	return handler;	
}

function getShelfObj(obj) {
	var shelfObj = {
		id: obj.id,
		name: obj.name,
		book_count: obj.book_count,
	};

	return shelfObj;
}
