// goodreads/node_modules/get-shelf.js
var https = require("https");
var xml2js = require("xml2js");

var config = {
	host: "www.goodreads.com",
	port: 443,
	key: "n8hATXd6wzepLEZo5zlhkg",
	user: "4175880-retorick"
};

exports.getShelf = function(req, res) {
	var shelf = req.params.shelf;

	grshelf(res, shelf).end();
}

function grshelf(res, shelf) {
	var parser = xml2js.Parser();

	var options = {
		hostname: config.host,
		port: config.port,
		path: "/review/list.xml?id=" + config.user + "&key=" + config.key + "&shelf=" + shelf,
		method: "GET"
	};
	
	var handler = 
		https.request(options, function(httpRes) {
			var body = "";

			httpRes.on("data", function(piece) {
				body += piece;
			});

			httpRes.on("end", function() {
				parser.parseString(body, function(err, data) {
					var books = data.GoodreadsResponse.books[0].book;
					var bookList = [];
					books.forEach(function(obj) {
						var bookObj = getBookObj(obj);

						bookList.push(bookObj);
					});
					res.json(bookList);
				});
			});

		});

	return handler;
}

function getBookObj(obj) {
	var authorArr = obj.authors[0].author.map(function(authObj) { return authObj.name[0]; });
	var author = authorArr.join(", ");	
	var pages = obj.num_pages[0];
	var isbn = obj.isbn[0];
	var image = obj.image_url[0];
	var small_image = obj.small_image_url[0];
	var publisher = obj.publisher[0];
	var publication_year = obj.publication_year[0];
	var format = obj.format[0];
	var bookObj = {
		title: obj.title[0],
		author: author,
		pages: pages,
		isbn: isbn,
		image: image,
		small_image: small_image,
		publisher: publisher,
		publication_year: publication_year,
		format: format,
	};

	return bookObj;
}
